### PR TITLE
[Php74][Php80] Handle TypedPropertyRector + StringableForToStringRector

### LINKED ISSUE #6862: `StringableForToStringRector` + `TypedPropertyRector` + `string|null` = ERROR
# Bug Report

| Subject        | Details                                                         |
| :------------- | :---------------------------------------------------------------|
| Rector version |  0.12.7 (Rector f068fd98590a7ece4bee280ebb13224562b6cff3)             |

## Minimal PHP Code Causing Issue

https://getrector.org/demo/1ec5bb75-1ea0-6252-b46d-d772a8906bd8

`StringableForToStringRector` and `TypedPropertyRector` when combined are throwing an error.

If I run `TypedPropertyRector` by itself it outputs
```diff
    ---------- begin diff ----------
@@ @@
     /**
      * @ORM\Column(type="string", length=255, nullable=true)
      */
-    private $nickname;
+    private ?string $nickname = null;

     public function __toString()
     {
    ----------- end diff -----------
```

If I run `StringableForToStringRector` by itself it outputs
```diff
    ---------- begin diff ----------
@@ @@
 /**
  * @ORM\Entity(repositoryClass=RectorTestRepository::class)
  */
-class RectorTest
+class RectorTest implements \Stringable
 {
     /**
      * @ORM\Column(type="string", length=255, nullable=true)
@@ @@
      */
     private $nickname;

-    public function __toString()
+    public function __toString(): string
     {
         return $this->nickname;
     }
    ----------- end diff -----------
```

When combined, they just error

```
╰─❯ vendor/bin/rector process src/Entity/RectorTest.php --dry-run --debug
[parsing] src/Entity/RectorTest.php
[refactoring] src/Entity/RectorTest.php
    [applying] Rector\Php80\Rector\Class_\StringableForToStringRector
    [applying] Rector\Php74\Rector\Property\TypedPropertyRector
PHP Fatal error:  Uncaught PHPStan\ShouldNotHappenException: Cannot create PHPStan\Type\UnionType with: string, string|null in phar://<snip>/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/Type/UnionType.php:33
Stack trace:
#0 phar://<snip>/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/Type/UnionType.php(47): PHPStan\Type\UnionType::PHPStan\Type\{closure}()
#1 <snip>/vendor/rector/rector/rules/TypeDeclaration/TypeInferer/PropertyTypeInferer/GetterPropertyTypeInferer.php(97): PHPStan\Type\UnionType->__construct(Array)
#2 <snip>/vendor/rector/rector/rules/TypeDeclaration/TypeInferer/PropertyTypeInferer.php(101): Rector\TypeDeclaration\TypeInferer\PropertyTypeInferer\GetterPropertyTypeInferer->inferProperty(Object(PhpParser\Node\Stmt\Property))
#3 <snip>/vendor/rector/rector/rules/TypeDeclaration/TypeInferer/PropertyTypeInferer.php(73): Rector\TypeDeclaration\TypeInferer\PropertyTypeInferer->getResolvedTypes(Object(PhpParser\Node\Stmt\Property))
#4 <snip>/vendor/rector/rector/rules/Php74/Rector/Property/TypedPropertyRector.php(172): Rector\TypeDeclaration\TypeInferer\PropertyTypeInferer->inferProperty(Object(PhpParser\Node\Stmt\Property))
#5 <snip>/vendor/rector/rector/src/Rector/AbstractRector.php(237): Rector\Php74\Rector\Property\TypedPropertyRector->refactor(Object(PhpParser\Node\Stmt\Property))
#6 <snip>/vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(176): Rector\Core\Rector\AbstractRector->enterNode(Object(PhpParser\Node\Stmt\Property))
#7 <snip>/vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(105): PhpParser\NodeTraverser->traverseArray(Array)
#8 <snip>/vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(196): PhpParser\NodeTraverser->traverseNode(Object(PhpParser\Node\Stmt\Class_))
#9 <snip>/vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(105): PhpParser\NodeTraverser->traverseArray(Array)
#10 <snip>/vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(196): PhpParser\NodeTraverser->traverseNode(Object(PhpParser\Node\Stmt\Namespace_))
#11 <snip>/vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(85): PhpParser\NodeTraverser->traverseArray(Array)
#12 <snip>/vendor/rector/rector/src/PhpParser/NodeTraverser/RectorNodeTraverser.php(56): PhpParser\NodeTraverser->traverse(Array)
#13 <snip>/vendor/rector/rector/src/Application/FileProcessor.php(57): Rector\Core\PhpParser\NodeTraverser\RectorNodeTraverser->traverse(Array)
#14 <snip>/vendor/rector/rector/src/Application/FileProcessor/PhpFileProcessor.php(143): Rector\Core\Application\FileProcessor->refactor(Object(Rector\Core\ValueObject\Application\File), Object(Rector\Core\ValueObject\Configuration))
#15 <snip>/vendor/rector/rector/src/Application/FileProcessor/PhpFileProcessor.php(103): Rector\Core\Application\FileProcessor\PhpFileProcessor->refactorNodesWithRectors(Object(Rector\Core\ValueObject\Application\File), Object(Rector\Core\ValueObject\Configuration))
#16 <snip>/vendor/rector/rector/src/Application/ApplicationFileProcessor.php(96): Rector\Core\Application\FileProcessor\PhpFileProcessor->process(Object(Rector\Core\ValueObject\Application\File), Object(Rector\Core\ValueObject\Configuration))
#17 <snip>/vendor/rector/rector/src/Application/ApplicationFileProcessor.php(74): Rector\Core\Application\ApplicationFileProcessor->processFiles(Array, Object(Rector\Core\ValueObject\Configuration))
#18 <snip>/vendor/rector/rector/src/Console/Command/ProcessCommand.php(165): Rector\Core\Application\ApplicationFileProcessor->run(Array, Object(Rector\Core\ValueObject\Configuration))
#19 <snip>/vendor/rector/rector/vendor/symfony/console/Command/Command.php(296): Rector\Core\Console\Command\ProcessCommand->execute(Object(RectorPrefix20211209\Symfony\Component\Console\Input\ArgvInput), Object(RectorPrefix20211209\Symfony\Component\Console\Output\ConsoleOutput))
#20 <snip>/vendor/rector/rector/vendor/symfony/console/Application.php(897): RectorPrefix20211209\Symfony\Component\Console\Command\Command->run(Object(RectorPrefix20211209\Symfony\Component\Console\Input\ArgvInput), Object(RectorPrefix20211209\Symfony\Component\Console\Output\ConsoleOutput))
#21 <snip>/vendor/rector/rector/vendor/symfony/console/Application.php(300): RectorPrefix20211209\Symfony\Component\Console\Application->doRunCommand(Object(Rector\Core\Console\Command\ProcessCommand), Object(RectorPrefix20211209\Symfony\Component\Console\Input\ArgvInput), Object(RectorPrefix20211209\Symfony\Component\Console\Output\ConsoleOutput))
#22 <snip>/vendor/rector/rector/src/Console/ConsoleApplication.php(71): RectorPrefix20211209\Symfony\Component\Console\Application->doRun(Object(RectorPrefix20211209\Symfony\Component\Console\Input\ArgvInput), Object(RectorPrefix20211209\Symfony\Component\Console\Output\ConsoleOutput))
#23 <snip>/vendor/rector/rector/vendor/symfony/console/Application.php(196): Rector\Core\Console\ConsoleApplication->doRun(Object(RectorPrefix20211209\Symfony\Component\Console\Input\ArgvInput), Object(RectorPrefix20211209\Symfony\Component\Console\Output\ConsoleOutput))
#24 <snip>/vendor/rector/rector/bin/rector.php(56): RectorPrefix20211209\Symfony\Component\Console\Application->run()
#25 <snip>/vendor/rector/rector/bin/rector(5): require_once('/Users/kopinsky...')
#26 {main}
  thrown in phar://<snip>/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/Type/UnionType.php on line 33
Fatal error: Uncaught PHPStan\ShouldNotHappenException: Cannot create PHPStan\Type\UnionType with: string, string|null in phar://<snip>/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/Type/UnionType.php:33
Stack trace:
#0 phar://<snip>/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/Type/UnionType.php(47): PHPStan\Type\UnionType::PHPStan\Type\{closure}()
#1 <snip>/vendor/rector/rector/rules/TypeDeclaration/TypeInferer/PropertyTypeInferer/GetterPropertyTypeInferer.php(97): PHPStan\Type\UnionType->__construct(Array)
#2 <snip>/vendor/rector/rector/rules/TypeDeclaration/TypeInferer/PropertyTypeInferer.php(101): Rector\TypeDeclaration\TypeInferer\PropertyTypeInferer\GetterPropertyTypeInferer->inferProperty(Object(PhpParser\Node\Stmt\Property))
#3 <snip>/vendor/rector/rector/rules/TypeDeclaration/TypeInferer/PropertyTypeInferer.php(73): Rector\TypeDeclaration\TypeInferer\PropertyTypeInferer->getResolvedTypes(Object(PhpParser\Node\Stmt\Property))
#4 <snip>/vendor/rector/rector/rules/Php74/Rector/Property/TypedPropertyRector.php(172): Rector\TypeDeclaration\TypeInferer\PropertyTypeInferer->inferProperty(Object(PhpParser\Node\Stmt\Property))
#5 <snip>/vendor/rector/rector/src/Rector/AbstractRector.php(237): Rector\Php74\Rector\Property\TypedPropertyRector->refactor(Object(PhpParser\Node\Stmt\Property))
#6 <snip>/vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(176): Rector\Core\Rector\AbstractRector->enterNode(Object(PhpParser\Node\Stmt\Property))
#7 <snip>/vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(105): PhpParser\NodeTraverser->traverseArray(Array)
#8 <snip>/vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(196): PhpParser\NodeTraverser->traverseNode(Object(PhpParser\Node\Stmt\Class_))
#9 <snip>/vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(105): PhpParser\NodeTraverser->traverseArray(Array)
#10 <snip>/vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(196): PhpParser\NodeTraverser->traverseNode(Object(PhpParser\Node\Stmt\Namespace_))
#11 <snip>/vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(85): PhpParser\NodeTraverser->traverseArray(Array)
#12 <snip>/vendor/rector/rector/src/PhpParser/NodeTraverser/RectorNodeTraverser.php(56): PhpParser\NodeTraverser->traverse(Array)
#13 <snip>/vendor/rector/rector/src/Application/FileProcessor.php(57): Rector\Core\PhpParser\NodeTraverser\RectorNodeTraverser->traverse(Array)
#14 <snip>/vendor/rector/rector/src/Application/FileProcessor/PhpFileProcessor.php(143): Rector\Core\Application\FileProcessor->refactor(Object(Rector\Core\ValueObject\Application\File), Object(Rector\Core\ValueObject\Configuration))
#15 <snip>/vendor/rector/rector/src/Application/FileProcessor/PhpFileProcessor.php(103): Rector\Core\Application\FileProcessor\PhpFileProcessor->refactorNodesWithRectors(Object(Rector\Core\ValueObject\Application\File), Object(Rector\Core\ValueObject\Configuration))
#16 <snip>/vendor/rector/rector/src/Application/ApplicationFileProcessor.php(96): Rector\Core\Application\FileProcessor\PhpFileProcessor->process(Object(Rector\Core\ValueObject\Application\File), Object(Rector\Core\ValueObject\Configuration))
#17 <snip>/vendor/rector/rector/src/Application/ApplicationFileProcessor.php(74): Rector\Core\Application\ApplicationFileProcessor->processFiles(Array, Object(Rector\Core\ValueObject\Configuration))
#18 <snip>/vendor/rector/rector/src/Console/Command/ProcessCommand.php(165): Rector\Core\Application\ApplicationFileProcessor->run(Array, Object(Rector\Core\ValueObject\Configuration))
#19 <snip>/vendor/rector/rector/vendor/symfony/console/Command/Command.php(296): Rector\Core\Console\Command\ProcessCommand->execute(Object(RectorPrefix20211209\Symfony\Component\Console\Input\ArgvInput), Object(RectorPrefix20211209\Symfony\Component\Console\Output\ConsoleOutput))
#20 <snip>/vendor/rector/rector/vendor/symfony/console/Application.php(897): RectorPrefix20211209\Symfony\Component\Console\Command\Command->run(Object(RectorPrefix20211209\Symfony\Component\Console\Input\ArgvInput), Object(RectorPrefix20211209\Symfony\Component\Console\Output\ConsoleOutput))
#21 <snip>/vendor/rector/rector/vendor/symfony/console/Application.php(300): RectorPrefix20211209\Symfony\Component\Console\Application->doRunCommand(Object(Rector\Core\Console\Command\ProcessCommand), Object(RectorPrefix20211209\Symfony\Component\Console\Input\ArgvInput), Object(RectorPrefix20211209\Symfony\Component\Console\Output\ConsoleOutput))
#22 <snip>/vendor/rector/rector/src/Console/ConsoleApplication.php(71): RectorPrefix20211209\Symfony\Component\Console\Application->doRun(Object(RectorPrefix20211209\Symfony\Component\Console\Input\ArgvInput), Object(RectorPrefix20211209\Symfony\Component\Console\Output\ConsoleOutput))
#23 <snip>/vendor/rector/rector/vendor/symfony/console/Application.php(196): Rector\Core\Console\ConsoleApplication->doRun(Object(RectorPrefix20211209\Symfony\Component\Console\Input\ArgvInput), Object(RectorPrefix20211209\Symfony\Component\Console\Output\ConsoleOutput))
#24 <snip>/vendor/rector/rector/bin/rector.php(56): RectorPrefix20211209\Symfony\Component\Console\Application->run()
#25 <snip>/vendor/rector/rector/bin/rector(5): require_once('/Users/kopinsky...')
#26 {main}
  thrown in phar://<snip>/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/Type/UnionType.php on line 33
```

## Expected Behaviour

Honestly, I'm not familiar enough with PHP8 to say what the right answer is, when the `__toString` returns something that is technically nullable. Maybe it should just skip it? But erroring certainly isn't the right answer ;-)

### PR DESCRIPTION
**Pull request by samsonasik**
Given the following code:

```php
declare(strict_types=1);

namespace Rector\Core\Tests\Issues\StringableTypedProperty\Fixture;

class Fixture
{
    private $nickname;

    public function __toString()
    {
        return $this->nickname;
    }

    public function getNickname(): ?string
    {
        return $this->nickname;
    }
}
```

It produce error:

```bash
1) Rector\Core\Tests\Issues\StringableTypedProperty\StringableTypedPropertyTest::test with data set #0 (Symplify\SmartFileSystem\SmartFileInfo Object (...))
PHPStan\ShouldNotHappenException: Cannot create PHPStan\Type\UnionType with: string, string|null
```

This PR fix it. Fixes https://github.com/rectorphp/rector/issues/6862

### CODE REVIEW

**Comment by samsonasik** [comment]
Fixed 🎉 The solution is only apply typed property to make nullable: 

```diff
-    private $nickname;
+    private ?string $nickname = null;
```

But not apply `implements Stringable` as the type returned to `__toString()` is not string.

**Comment by samsonasik** [comment]
@mkopinsky it seems in PHP 8.0, when class has method `__toString()`, it must return string, without return string, it will got error:

```bash
Fatal error: Uncaught TypeError: Fixture::__toString(): Return value must be of type string, null returned in /in/dhmYZ:7
Stack trace:
#0 /in/dhmYZ(11): Fixture->__toString()
```

ref https://3v4l.org/dhmYZ

**Comment by samsonasik** [comment]
It seems the solution is find return statement in the classmethod with `__toString()` method, if not found, return empty string '', if found and not string, cast to (string), eg:

```diff
-return $this->nickname;
+return (string) $this->nickname;
```

**Comment by mkopinsky** [comment]
That makes perfect sense.

In the application I took this from, `$this->nickname` in practice is never null, which is why I haven't hit that fatal error. I guess the other alternative (which is beyond the scope of rector, but is an option for me) is to actually change the column (via the annotation) to be not null.

**Comment by samsonasik** [comment]
@mkopinsky implemented 🎉

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba it is ready for review.

**Comment by TomasVotruba** [comment]
Thank you, looks good :+1:

File: rules/Php80/Rector/Class_/StringableForToStringRector.php [hunk 1]
```diff
@@ -106,4 +118,38 @@ public function refactor(Node $node): ?Node
 
         return $node;
     }
+
+    private function processNotStringType(ClassMethod $toStringClassMethod): void
+    {
+        $hasReturn = $this->betterNodeFinder->hasInstancesOfInFunctionLikeScoped($toStringClassMethod, Return_::class);
+
+        if (! $hasReturn) {
+            $lastKey = array_key_last((array) $toStringClassMethod->stmts);
```
Comment (samsonasik) [inline rules/Php80/Rector/Class_/StringableForToStringRector.php:127]: it seems cause downgrade error:

https://github.com/rectorphp/rector-src/runs/4505677100?check_suite_focus=true#step:14:80

```diff
-            $lastKey = array_key_last((array) $toStringClassMethod->stmts);
+            end((array) $toStringClassMethod->stmts);
+            $lastKey = key((array) $toStringClassMethod->stmts);
```

`DowngradeArrayKeyFirstLastRector` seems needs update for casting condition.

Comment (samsonasik) [reply rules/Php80/Rector/Class_/StringableForToStringRector.php:127]: I created PR https://github.com/rectorphp/rector-src/pull/1475 for it. It next need a tweak in `DowngradeArrayKeyFirstLastRector` to adapt this condition.

Comment (TomasVotruba) [reply rules/Php80/Rector/Class_/StringableForToStringRector.php:127]: Thanks :+1:

**Pull request merged**